packages/expo-media-library/ios/EXMediaLibrary/EXMediaLibraryImageLoader.m

```diff
@@ -0,0 +1,110 @@
+#if __has_include(<React/RCTImageURLLoader.h>)
+
+#import <Photos/Photos.h>
+#import <React/RCTUtils.h>
+#import <EXMediaLibrary/EXMediaLibraryImageLoader.h>
+
+@implementation EXMediaLibraryImageLoader
+
+RCT_EXPORT_MODULE()
+
+#pragma mark - RCTImageURLLoader
+
+- (BOOL)canLoadImageURL:(NSURL *)requestURL
+{
+  if (![PHAsset class]) {
+    return NO;
+  }
+  return [requestURL.scheme caseInsensitiveCompare:@"assets-library"] == NSOrderedSame ||
+    [requestURL.scheme caseInsensitiveCompare:@"ph"] == NSOrderedSame;
+}
+
+- (RCTImageLoaderCancellationBlock)loadImageForURL:(NSURL *)imageURL
+                                              size:(CGSize)size
+                                             scale:(CGFloat)scale
+                                        resizeMode:(RCTResizeMode)resizeMode
+                                   progressHandler:(RCTImageLoaderProgressBlock)progressHandler
+                                partialLoadHandler:(RCTImageLoaderPartialLoadBlock)partialLoadHandler
+                                 completionHandler:(RCTImageLoaderCompletionBlock)completionHandler
+{
+  // Using PhotoKit for iOS 8+
+  // The 'ph://' prefix is used by FBMediaKit to differentiate between
+  // assets-library. It is prepended to the local ID so that it is in the
+  // form of an NSURL which is what assets-library uses.
+  NSString *assetID = @"";
+  PHFetchResult *results;
+  if (!imageURL) {
+    completionHandler(RCTErrorWithMessage(@"Cannot load a photo library asset with no URL"), nil);
+    return ^{};
+  } else if ([imageURL.scheme caseInsensitiveCompare:@"assets-library"] == NSOrderedSame) {
+    assetID = [imageURL absoluteString];
+    results = [PHAsset fetchAssetsWithALAssetURLs:@[imageURL] options:nil];
+  } else {
+    assetID = [imageURL.absoluteString substringFromIndex:@"ph://".length];
+    results = [PHAsset fetchAssetsWithLocalIdentifiers:@[assetID] options:nil];
+  }
+  if (results.count == 0) {
+    NSString *errorText = [NSString stringWithFormat:@"Failed to fetch PHAsset with local identifier %@ with no error message.", assetID];
+    completionHandler(RCTErrorWithMessage(errorText), nil);
+    return ^{};
+  }
+
+  PHAsset *asset = [results firstObject];
+  PHImageRequestOptions *imageOptions = [PHImageRequestOptions new];
+
+  // Allow PhotoKit to fetch images from iCloud
+  imageOptions.networkAccessAllowed = YES;
+
+  if (progressHandler) {
+    imageOptions.progressHandler = ^(double progress, NSError *error, BOOL *stop, NSDictionary<NSString *, id> *info) {
+      static const double multiplier = 1e6;
+      progressHandler(progress * multiplier, multiplier);
+    };
+  }
+
+  // Note: PhotoKit defaults to a deliveryMode of PHImageRequestOptionsDeliveryModeOpportunistic
+  // which means it may call back multiple times - we probably don't want that
+  imageOptions.deliveryMode = PHImageRequestOptionsDeliveryModeHighQualityFormat;
```

Comment:
Does this mean that if a user attempts to upload a `ph://` URI directly using `fetch`, the image might get upscaled (increase in size)? In stock React Native there’s the `ph-upload://` to be used in uploads to avoid the upscaling issue

Reply:
Yeah, I think you may be right 🤔 My goal was to make it possible to render images with such urls and generally this is the same as photo library image loader from RN: [RCTPhotoLibraryImageLoader.mm](https://github.com/facebook/react-native/blob/master/Libraries/CameraRoll/RCTPhotoLibraryImageLoader.mm).
However, looks like it doesn't handle `ph-upload://` urls and for that case we would need to copy [RCTAssetsLibraryRequestHandler.mm](https://github.com/facebook/react-native/blob/master/Libraries/CameraRoll/RCTAssetsLibraryRequestHandler.mm) as well. They both provide similar functionality and they both fix an issue with rendering, so I thought we don't need both of them at once. Also looks like stock React Native doesn't include them by default since this is the old code of CameraRoll from before it's been moved to separate repo.

Is this thing important to you? Would you be able to open similar PR but copying request handler functionality?

Reply:
It’s not important to me personally, I just figured I’d mention it :)

Reply:
Ok, cool, I'll try to fix this anyway before the release 😉 Thanks for pointing this out! 🙇‍♂️